ingest/scripts/extract-genbank-annotations.py

```diff
@@ -0,0 +1,332 @@
+        return {"epiweek": "", "epiyear": str(_normalize_year(match.group(1)))}
+
+    return {"epiweek": "", "epiyear": ""}
+
+
+def _parse_genotype(maybe_genotype):
+    """
+    Given a string that may contain a rubella genotype, extract the
+    genotype, normalize it to accepted WHO format, and return it (or
+    return the empty string if no genotype is found).
+    """
+    match = re.search(r"([12][aAbBcCdDeEfFgGhHiIjJ])", maybe_genotype)
+
+    if match:
+        genotype = match.group(1)
+        if re.search(r"1[aA]", genotype):
+            return genotype.lower()
+        else:
+            return genotype.upper()
+
+    return ""
+
+
+def _parse_source_feature(qualifiers):
+    """
+    Given a set of qualifiers from the "source" feature of a GenBank
+    record, attempt to extract and return the "collection_date",
+    "epiweek" and "epiyear" extracted from WHO strain name, and
+    "genotype" extracted also extracted from the WHO strain name.
+
+    Results are returned as a dict with keys named above; if values
+    cannot be extracted, the dict will have empty string values.
+    """
+    results = {
+        "collection_date": "",
+        "epiweek": "",
+        "epiyear": "",
+        "genotype": "",
+    }
+
+    # NOTE: SeqIO.parse turns all the source features into single element
+    # lists, which is why all the `.get()` calls below append a `[0]`
+
+    # extract the collection date, if it's present
+    results["collection_date"] = qualifiers.get("collection_date", [""])[0]
+
+    # Try to parse genotype and epidate, first from `strain`, and then
+    # from `isolate` only if `strain` is not present. This ordering is
+    # because, from empirical inspection, when both are present,
+    # strain contains the WHO strain designation and isolate does not
+    # — but when strain is missing, sometimes isolate has the WHO
+    # strain name.
+    isolate = qualifiers.get("isolate", [""])[0]
+    strain = qualifiers.get("strain", [""])[0]
+
+    if strain:
+        _parse_strain_name(strain, results)
+    elif isolate:
+        _parse_strain_name(isolate, results)
+
+    # if we didn't find a genotype in the strain or isolate, see if
+    # it's in the note
+    if not results["genotype"]:
+        note = qualifiers.get("note", [""])[0]
+        results["genotype"] = _parse_genotype(note)
+
+    return results
+
+
+def _parse_strain_name(name, results):
+    """
+    Given a putative WHO strain name and a results dict, attempt to
+    extract epiweek, epiyear, and genotype from the strain name, and
+    update the provided results dict accordingly.
+    """
+    if re.match(r"^RV[is]/", name):
+        parts = name.split("/", 3)[1:]
+
+        if len(parts) > 2:
+            location, date, rest = parts
+
+            parsed_date = _parse_date(date)
+            results["epiweek"] = parsed_date["epiweek"]
+            results["epiyear"] = parsed_date["epiyear"]
+
+            results["genotype"] = _parse_genotype(rest)
```

Comment:
Just realized this doesn't support one of the example strains in your docs at the top (`RVs/Rouen.FRA/97[1E]`). Should this include path for `elif len(parts) == 2`?

I only noticed because I saw [AB745031](https://www.ncbi.nlm.nih.gov/nuccore/AB745031) had a fully masked date even though the strain has the epiweek date (`RVs/HoChiMinh.VNM/40.11[2B]`)

Reply:
hmm, will take a deeper look, thanks.